### ISSUE TITLE
Memory leak when upgrading from tcp frontend to http backend

### ISSUE THREAD

**Issue by chtsanti** [haproxy/haproxy#935]
## Detailed description of the problem

The haproxy2.1 is used however the bug exists also in haproxy2.2  and in haproxy master branch.

A very simple configuration is used, with a simple tcp frontend which forwards to a http backend.
The command:
     ```# echo "show pools" | socat - /tmp/haproxy_admin.sock```
 shows that a pool, in my case the "email_alert", is constantly growing. In the production installation the "tcpcheck_ru" is the pool which is growing.

## Expected behavior

The haproxy memory should not grow infinitely.

## Steps to reproduce the behavior

Sending some thousands of requests on a haproxy with simple setup with tcp frontend which forwards to an http backend is enough to reproduce the issue.

## Do you have any idea what may have caused this?

I assume latest haproxy2.1 source code. Here are the sequence which is called:

  * The tcp frontend forwards to an HTTP backend. 
  * The stream_set_backend function is called (proxy.c line 1404)
  * The conn_upgrade_mux_fe (include/proto/connection.h, line 1158) called to upgrade the client mux.
  * A new mux is build and the old one is destroyed with a mux_pt_destroy
    (mux_pt.c, line 27) function call. 
   * Inside the mux_pt_destroy function the the ctx->wait_event.tasklet is never released  because of the
    "ctx->conn->ctx == ctx" check which fails.

## Do you have an idea how to solve the issue?

I have a simple branch which fixes the issue for us:
   https://github.com/chtsanti/haproxy/tree/haproxy-tasklet-leak

## What is your configuration?

```
frontend k_proxy
         bind :9080 
         mode tcp
         timeout client  10m
         default_backend k_squid_bump

backend k_squid_bump
        mode http
        timeout connect 10s
        timeout server  60s
        server s3 127.0.0.1:8080
```

## Output of `haproxy -vv` and `uname -a`

```
# haproxy -vv
Sharing hlua with ssl_sock_ct
Sharing fcgi_strm with h2s
Sharing uniqueid with ssl_sock_ct
Sharing h1c with http_txn
Sharing spoe_appctx with h2s
Sharing tcpcheck_rule with email_alert
Sharing dns_requester with email_alert
Sharing session server list with email_alert
Sharing fcgi_flt_ctx with cache_st
Sharing fcgi_hdr_rule with fcgi_param_
Sharing connection with h1s
Sharing conn_stream with email_alert
Sharing sockaddr with ssl_sock_ct
Sharing authority with session
Sharing filter with email_alert
Sharing task with http_txn
Sharing tasklet with email_alert
Sharing notification with fcgi_param_
Sharing comp_state with cache_st
Sharing vars with fcgi_param_
Sharing mux_pt with cache_st
Sharing pendconn with fcgi_param_
Sharing xprt_handshake_ctx_pool with email_alert
Sharing appctx with h2s
Sharing sig_handlers with cache_st
Sharing comp_ctx with http_txn
Sharing pipe with cache_st
HA-Proxy version 2.1.4-techloq-8d2cbd-275 2020/10/27 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.4.html
Running on: Linux 5.4.0-51-generic #56~18.04.1-Ubuntu SMP Tue Oct 6 09:47:18 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=12).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [TRACE] trace
        [COMP] compression
```

```
# uname -a
Linux xxxxxxxx 5.4.0-51-generic #56~18.04.1-Ubuntu SMP Tue Oct 6 09:47:18 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by capflam** [comment]
I confirm the leak. I will fix it soon. Thanks !

**Comment by capflam** [comment]
Should be fixed now in 2.3-dev.

**Comment by capflam** [comment]
Now backported in all versions. Thanks !
